File: cpu-monitor.py
#!/usr/bin/env python3

# TRIX: Inference of Transient Violation Times from Logged Routing Events or Collected BGP Messages
# Copyright (C) 2024-2025 Roland Schmid <roschmi@ethz.ch> and Tibor Schneider <sctibor@ethz.ch>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import psutil
import argparse
import time
import os
from datetime import datetime
import csv


def read_control(control_file):
    """Read the control file and return its integer content (1 or 0)."""
    try:
        with open(control_file, "r") as f:
            return int(f.read().strip())
    except (FileNotFoundError, ValueError):
        return 0


def monitor_all_processes(
    interval, control_file, router_id, router_name="", output_file=None
):
    """Monitor all processes. Data collection continues until control_file contains something else than a 1."""
    data_log = []

    # wait before starting the measurements
    while read_control(control_file) != 1:
        time.sleep(interval)

    attrs = [
        "pid",
        "name",
        "exe",
        "cpu_percent",
        "memory_percent",
        "num_threads",
        "nice",
        "num_fds",
        "open_files",
    ]
    headers = ["rid", "router_name", "timestamp"] + attrs

    cont = True
    while cont:
        for proc in psutil.process_iter(attrs):
            timestamp = datetime.utcnow().timestamp()
            row = (router_id, router_name, timestamp) + tuple(
                proc.info[attr] for attr in attrs
            )
            data_log.append(row)

            time.sleep(interval)
            cont = read_control(control_file) == 1
            if not cont:
                break

    if output_file:
        # Write the collected data to a CSV file
        os.makedirs(os.path.dirname(output_file), exist_ok=True)
        with open(output_file, mode="w", newline="") as file:
            writer = csv.writer(file)
            writer.writerow(headers)
            for entry in data_log:
                writer.writerow(entry)
    else:
        print(",".join(headers))
        for entry in data_log:
            print(",".join([str(x) for x in entry]))


def monitor_bgp_cpu(
    interval, control_file, router_id, router_name="", output_file=None
):
    """Monitor system and the processes relevant for BGP workloads for CPU usage and log results to a file. Data collection continues until control_file contains something else than a 1."""
    data_log = []

    # find the BGP process
    for bgp_process in psutil.process_iter():
        if bgp_process.name() == "bgp":
            break
    if not bgp_process:
        print("BGP process not found.")
        return

    # find the ipfib process
    for ipfib_process in psutil.process_iter():
        if ipfib_process.name() == "ipfib":
            break
    if not ipfib_process:
        print("'ipfib' process not found.")
        return

    # find the urib process
    for urib_process in psutil.process_iter():
        if urib_process.name() == "urib":
            break
    if not urib_process:
        print("'urib' process not found.")
        return

    # wait before starting the measurements
    while read_control(control_file) != 1:
        time.sleep(interval)

    # reset counters
    last_cpu = psutil.cpu_percent(percpu=True)
    last_bgp_cpu = bgp_process.cpu_percent()
    last_ipfib_cpu = ipfib_process.cpu_percent()
    last_urib_cpu = urib_process.cpu_percent()

    # get number of CPUs to define columns dynamically
    num_cpus = len(last_cpu)
    cpu_columns = [f"cpu{i+1}" for i in range(num_cpus)]  # e.g., cpu1, cpu2, ..., cpuX
    headers = ["rid", "router_name", "timestamp", "cpu"] + cpu_columns + ["bgp_cpu", "ipfib_cpu", "urib_cpu"]

    while read_control(control_file) == 1:
        # collect data
        timestamp = datetime.utcnow().timestamp()
        system_cpu_percent = psutil.cpu_percent(percpu=True)
        bgp_process_cpu_percent = bgp_process.cpu_percent()
        ipfib_process_cpu_percent = ipfib_process.cpu_percent()
        urib_process_cpu_percent = urib_process.cpu_percent()

        total_cpu_percent = sum(system_cpu_percent)

        # store data
        data_log.append(
            (
                router_id,
                router_name,
                timestamp,
                total_cpu_percent,
                *system_cpu_percent,
                bgp_process_cpu_percent,
                ipfib_process_cpu_percent,
                urib_process_cpu_percent,
            )
        )

        # sleep for the specified interval
        time.sleep(interval)

    if output_file:
        # Write the collected data to a CSV file
        os.makedirs(os.path.dirname(output_file), exist_ok=True)
        with open(output_file, mode="w", newline="") as file:
            writer = csv.writer(file)
            writer.writerow(headers)
            for entry in data_log:
                writer.writerow(entry)
    else:
        print(",".join(headers))
        for entry in data_log:
            print(",".join([str(x) for x in entry]))


def main():
    parser = argparse.ArgumentParser(
        description="Monitor CPU usage of the BGP process."
    )
    parser.add_argument(
        "-r",
        "--router-id",
        type=int,
        required=True,
        help="RouterId of the monitored router.",
    )
    parser.add_argument(
        "-n",
        "--router-name",
        type=str,
        default="",
        help="Router name of the monitored router.",
    )
    parser.add_argument(
        "-i",
        "--interval",
        type=float,
        default=0.01,
        help="Time interval between measurements in seconds (default: 0.01)",
    )
    parser.add_argument(
        "-c",
        "--control-file",
        type=str,
        default=".router_lab_cpu_monitor_control",
        help="Path to the control file (default: .router_lab_cpu_monitor_control)",
    )
    parser.add_argument(
        "--std-out",
        action="store_true",
        help="Print results to stdout instead of writing to a file.",
    )
    parser.add_argument(
        "--all-processes", action="store_true", help="Get CPU usage of all processes."
    )
    args = parser.parse_args()

    if args.std_out:
        output_file = None
    else:
        # prepare output file path based on current timestamp
        timestamp = datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
        output_file = f"cpu_log/cpu_log_{timestamp}.csv"

    if args.all_processes:
        monitor_all_processes(
            args.interval,
            args.control_file,
            args.router_id,
            args.router_name,
            output_file,
        )
    else:
        # start basic CPU monitoring
        monitor_bgp_cpu(
            args.interval,
            args.control_file,
            args.router_id,
            args.router_name,
            output_file,
        )


if __name__ == "__main__":
    main()

File: src/analyzer/cpu_monitor.rs
// TRIX: Inference of Transient Violation Times from Logged Routing Events or Collected BGP Messages
// Copyright (C) 2024-2025 Roland Schmid <roschmi@ethz.ch> and Tibor Schneider <sctibor@ethz.ch>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//! Enables a CPU monitor on the cisco routers, tracking their load throughout experiments.

use std::{
    collections::HashMap,
    fs,
    path::{Path, PathBuf},
    process::Stdio,
};

use itertools::Itertools;
use tokio::{
    io::{AsyncReadExt, AsyncWriteExt},
    process::Child,
};

use bgpsim::prelude::*;
use router_lab::{ssh::SshSession, RouterLab};

use crate::{records::CpuRecord, Prefix};

const CPU_MONITOR_FILE: &str = ".router_lab_cpu_monitor.py";
const CPU_MONITOR_CONTROL_FILE: &str = ".router_lab_cpu_monitor_control";

async fn control_monitoring(
    ssh_name: impl AsRef<str>,
    enabled: bool,
) -> Result<(), Box<dyn std::error::Error>> {
    let _ = SshSession::new(ssh_name.as_ref())
        .await?
        .command(format!(
            "run bash echo {} > {CPU_MONITOR_CONTROL_FILE}",
            enabled as i32
        ))
        .output()
        .await?;

    Ok(())
}

/// Transfer the cpu monitor script to all routers, initialize control file and start the
/// monitoring processes.
pub async fn setup_cpu_monitoring<Q, Ospf: OspfImpl, S>(
    net: &Network<Prefix, Q>,
    lab: &RouterLab<'_, Prefix, Q, Ospf, S>,
) -> Result<HashMap<RouterId, Child>, Box<dyn std::error::Error>> {
    let mut monitors = HashMap::new();

    for (rid, (router_properties, _)) in lab.routers() {
        // create the control file and disable the monitoring
        control_monitoring(&router_properties.ssh_name, false).await?;

        // transfer the cpu monitoring file and start the process
        // cat ../../cpu-monitor.py | ssh -T cisco-nexus1 "run bash cat > {FILE}"
        let session = SshSession::new(&router_properties.ssh_name).await?;
        let mut transfer = session
            .command("-T")
            .arg(format!("run bash cat > {CPU_MONITOR_FILE}"))
            .stdin(Stdio::piped())
            .spawn()?;
        transfer
            .stdin
            .take()
            .unwrap()
            .write_all(include_str!("../../cpu-monitor.py").as_bytes())
            .await?;
        transfer.wait().await?;

        // ssh cisco-nexus1 "run bash python3 cpu-monitor.py --std-out --router-id X --router-name Y"
        let child = session
            .command(format!(
                "run bash python3 {CPU_MONITOR_FILE} --control-file {CPU_MONITOR_CONTROL_FILE} --interval 0.1 --std-out --router-id {} --router-name {}",
                rid.index(),
                rid.fmt(net),
            ))
            .stdout(Stdio::piped())
            .spawn()?;

        monitors.insert(*rid, child);
    }

    Ok(monitors)
}

/// Start the extended monitoring processes. Has to be called after `setup_cpu_monitoring`. Will
/// automatically stop once `stop_cpu_monitoring` has been called, but should be `.await?`ed
/// nonetheless.
pub async fn setup_extended_cpu_monitoring<Q, Ospf: OspfImpl, S>(
    net: &Network<Prefix, Q>,
    lab: &RouterLab<'_, Prefix, Q, Ospf, S>,
    output_dir: impl AsRef<Path>,
) -> Result<HashMap<RouterId, Child>, Box<dyn std::error::Error>> {
    let output_dir = output_dir.as_ref();
    std::fs::create_dir_all(output_dir)?;
    let mut monitors = HashMap::new();

    for (rid, (router_properties, _)) in lab.routers() {
        let session = SshSession::new(&router_properties.ssh_name).await?;
        // ssh cisco-nexus1 "run bash python3 cpu-monitor.py --std-out --router-id X --router-name Y"
        let mut output_path = output_dir.to_path_buf();
        output_path.push(format!("{}.csv", rid.fmt(net)));

        let child = session
            .command(format!(
                "run bash python3 {CPU_MONITOR_FILE} --control-file {CPU_MONITOR_CONTROL_FILE} --interval 0.001 --std-out --router-id {} --router-name {} --all-processes",
                rid.index(),
                rid.fmt(net),
            ))
            .stdout(fs::OpenOptions::new()
                    .write(true)
                    .create(true)
                    .truncate(true)
                    .open(&output_path)
                    .unwrap()
                )
            .spawn()?;

        monitors.insert(*rid, child);
    }

    Ok(monitors)
}

/// Activate the (running) monitoring processes. Requires that `setup_cpu_monitoring` has
/// successfully been called previously.
pub async fn start_cpu_monitoring<Q, Ospf: OspfImpl, S>(
    lab: &RouterLab<'_, Prefix, Q, Ospf, S>,
) -> Result<(), Box<dyn std::error::Error>> {
    for (router_properties, _) in lab.routers().values() {
        // start the monitoring
        control_monitoring(&router_properties.ssh_name, true).await?;
    }
    Ok(())
}

/// Stop the ongoing CPU monitoring and collects all data into a single csv file.
pub async fn stop_cpu_monitoring<Q, Ospf: OspfImpl, S>(
    lab: &RouterLab<'_, Prefix, Q, Ospf, S>,
    monitors: HashMap<RouterId, Child>,
    output_path: impl AsRef<Path>,
) -> Result<PathBuf, Box<dyn std::error::Error>> {
    let mut records = Vec::new();

    // stop the monitoring, capture the outputs into a buffer, and combine all records into a Vec
    for (rid, mut child) in monitors.into_iter() {
        // disable the monitoring
        control_monitoring(&lab.routers().get(&rid).unwrap().0.ssh_name, false).await?;

        let mut buffer = Vec::new();
        let mut stdout = child.stdout.take().unwrap();
        stdout.read_to_end(&mut buffer).await?;

        // deserialize into records vec
        let mut csv = csv::Reader::from_reader(buffer.as_slice());
        for record in csv.deserialize() {
            let record: CpuRecord = record.unwrap();
            records.push(record);
        }
    }

    // sort & write combined csv
    let output_path = output_path.as_ref().to_path_buf();
    let mut csv = csv::WriterBuilder::new().has_headers(true).from_writer(
        fs::OpenOptions::new()
            .write(true)
            .create(true)
            .truncate(true)
            .open(&output_path)
            .unwrap(),
    );
    for record in records
        .into_iter()
        .sorted_by(|a, b| a.timestamp.total_cmp(&b.timestamp))
    {
        csv.serialize(record).unwrap();
    }
    csv.flush().unwrap();

    Ok(output_path)
}
